### PR TITLE
Fix #37: standardize error handling and call the C destructors

### LINKED ISSUE #37: Exceptions: Mismatched free() / delete / delete []
# Description

Controller actions receive a C result structure allocated with malloc.

If the response status is 200, the result is copied in a C++ structure, and the C struct is freed using its dedicated method.

If an error occured, the C struct is passed over to `throwExceptionFromStatus`. Here, there are several problems accumulating:

1. While `result->stack` is freed with the C `free` function, the `result->error` is freed with `delete`. Only memory allocated with `new` should be freed with `delete`
2. Just before throwing the exception, the struct itself is freed with `delete` instead of `free`
3. none of the struct properties are freed. While this is not a problem most of the time, it is if a [PartialError](https://docs-v2.kuzzle.io/api/1/essentials/errors/#specific-errors-default) is returned by Kuzzle since, in that case, it means that the `result` part is not NULL

So we have a free/delete discrepancy along with possible memory leaks whenever Kuzzle answers with an error message.

# Corresponding Valgrind report

```
Mismatched free() / delete / delete []
   at 0x4C3123B: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x7F7ED7: void kuzzleio::throwExceptionFromStatus<kuzzleio::string_result>(kuzzleio::string_result*) (exceptions.hpp:102)
   by 0x80D8D4: kuzzleio::Document::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, kuzzleio::s_query_options*) (document.cpp:67)
   by 0x797798: (anonymous namespace)::CukeObject6::bodyWithArgs() (collection-steps.cpp:84)
   by 0x79BDB0: void cucumber::internal::BasicStep::invokeWithIndexedArgs<(anonymous namespace)::CukeObject6, void>((anonymous namespace)::CukeObject6&, void ((anonymous namespace)::CukeObject6::*)(), cucumber::internal::index_sequence<>) (StepManager.hpp:142)
   by 0x799BEE: void cucumber::internal::BasicStep::invokeWithArgs<(anonymous namespace)::CukeObject6, void>((anonymous namespace)::CukeObject6&, void ((anonymous namespace)::CukeObject6::*)()) (StepManager.hpp:150)
   by 0x797693: (anonymous namespace)::CukeObject6::body() (collection-steps.cpp:75)
   by 0x82135B: void boost::_mfi::mf0<void, cucumber::internal::BasicStep>::call<cucumber::internal::BoostStep*>(cucumber::internal::BoostStep*&, void const*) const (in /home/scottinet/git/sdk-cpp/test/_build_cpp_tests/KuzzleSDKStepDefs)
   by 0x8212DA: void boost::_mfi::mf0<void, cucumber::internal::BasicStep>::operator()<cucumber::internal::BoostStep*>(cucumber::internal::BoostStep*&) const (in /home/scottinet/git/sdk-cpp/test/_build_cpp_tests/KuzzleSDKStepDefs)
   by 0x8210EF: void boost::_bi::list1<boost::_bi::value<cucumber::internal::BoostStep*> >::operator()<boost::_mfi::mf0<void, cucumber::internal::BasicStep>, boost::_bi::list0>(boost::_bi::type<void>, boost::_mfi::mf0<void, cucumber::internal::BasicStep>&, boost::_bi::list0&, int) (in /home/scottinet/git/sdk-cpp/test/_build_cpp_tests/KuzzleSDKStepDefs)
   by 0x820F29: boost::_bi::bind_t<void, boost::_mfi::mf0<void, cucumber::internal::BasicStep>, boost::_bi::list1<boost::_bi::value<cucumber::internal::BoostStep*> > >::operator()() (in /home/scottinet/git/sdk-cpp/test/_build_cpp_tests/KuzzleSDKStepDefs)
   by 0x820CE2: boost::detail::function::void_function_obj_invoker0<boost::_bi::bind_t<void, boost::_mfi::mf0<void, cucumber::internal::BasicStep>, boost::_bi::list1<boost::_bi::value<cucumber::internal::BoostStep*> > >, void>::invoke(boost::detail::function::function_buffer&) (in /home/scottinet/git/sdk-cpp/test/_build_cpp_tests/KuzzleSDKStepDefs)
 Address 0x9e95320 is 0 bytes inside a block of size 24 alloc'd
   at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x526FEA3: _cgo_585e479d2dd6_Cfunc__Cmalloc (_cgo_export.c:7625)
   by 0x500CED7: runtime.asmcgocall (/usr/lib/go-1.10/src/runtime/asm_amd64.s:688)
   by 0xC42000047F: ???
   by 0x1FFEFFD597: ???
   by 0x5008A63: runtime.(*mcache).nextFree.func1 (/usr/lib/go-1.10/src/runtime/malloc.go:556)
   by 0xC4200005FF: ???
   by 0x4FE69DF: ??? (/usr/lib/go-1.10/src/runtime/proc.go:1092)
   by 0x1FFEFFD5AF: ???
   by 0x1FFEFFD647: ???
   by 0x1FFEFFD5EF: ???
   by 0x500CF6D: runtime.cgocallback (/usr/lib/go-1.10/src/runtime/asm_amd64.s:737)
```

### PR DESCRIPTION
**Pull request by scottinet**
:warning: depends on https://github.com/kuzzleio/sdk-c/pull/46 :warning: 

# Description

This is a stabilization PR, aiming to fix some of the memory issues we have with this SDK:

* The static Kuzzle exception throwing method has been changed: it should not deallocate the memory of the structure provided to it. It now only throws the appropriate exception.
* A new `KUZZLE_API` macro has been added: it DRYifies and standardizes the way the Kuzzle API is requested, and how errors are handled. The main purpose of that macro is to correctly free the C result struct before throwing an exception
* All controller actions now use the new KUZZLE_API macro
* Fix some controller actions using `delete` to free the C result struct, instead of the corresponding C destructor

Note: there are other obvious memleaks remaining. I will address those in other, dedicated PRs, to make this stabilization pass easier to review.

# Other changes

* As per the changes made in the C SDK: the `user_rights_result` struct does not contain a NULL-terminated `result` pointer anymore, but a `rights` pointer, with its size stored in the `rights_length` property
* Fix non-standard indentation in a few files

### CODE REVIEW

File: src/auth.cpp [hunk 1]
```diff
@@ -38,91 +38,79 @@ namespace kuzzleio {
...
-    string_array_result *r = kuzzle_get_strategies(_auth, options);
-    if (r->error != nullptr)
-        throwExceptionFromStatus(r);
+    KUZZLE_API(string_array_result, r, kuzzle_get_strategies(_auth, options))
 
     std::vector<std::string> strategies;
 
-    for (size_t i = 0; r->result[i]; ++i)
+    for (size_t i = 0; i < r->result_length; ++i)
       strategies.push_back(r->result[i]);
```
Comment (Aschen) [inline src/auth.cpp]: You can initialize a `std::vector<std::string>` directly from a `char**` if you know the length:

```
std::vector<std::string> v = std::vector<std::string>(r->result, r->result + r->result_length)
```

File: src/index.cpp [hunk 2]
```diff
@@ -16,106 +16,93 @@
...
+  }
+
+  void Index::delete_(const std::string& index, query_options *options) {
+    KUZZLE_API(error_result, r, kuzzle_index_delete(_index, const_cast<char*>(index.c_str()), options))
+    kuzzle_free_error_result(r);
+  }
+
+  std::vector<std::string> Index::mDelete(const std::vector<std::string>& indexes, query_options *options) {
+    char **indexesArray = new char *[indexes.size()];
+    int i = 0;
```
Comment (Aschen) [inline src/index.cpp]: Useless variable creation in outer scope. A `size_t` could be used in the `for` for both array

File: src/document.cpp [hunk 3]
```diff
@@ -17,221 +17,189 @@
...
+    unregisterDocument(_document);
+    delete(_document);
+  }
 
-    std::string Document::create(const std::string& index, const std::string& collection, const std::string& id, const std::string& body, query_options *options) {
-        string_result *r = kuzzle_document_create(_document, const_cast<char*>(index.c_str()), const_cast<char*>(collection.c_str()), const_cast<char*>(id.c_str()), const_cast<char*>(body.c_str()), options);
-        if (r->error != nullptr)
-            throwExceptionFromStatus(r);
+  int Document::count(const std::string& index, const std::string& collection, const std::string& body, query_options *options) {
+    char* rbody = NULL;
```
Comment (Aschen) [inline src/document.cpp]:
```suggestion
    char* rbody = nullptr;
```

File: src/document.cpp [hunk 4]
```diff
@@ -17,221 +17,189 @@
...
-        std::string ret = r->result;
-        kuzzle_free_string_result(r);
-        return ret;
-    }
+  std::vector<std::string> Document::deleteByQuery(const std::string& index, const std::string& collection, const std::string& body, query_options *options) {
+    KUZZLE_API(string_array_result, r, kuzzle_document_delete_by_query(_document, const_cast<char*>(index.c_str()), const_cast<char*>(collection.c_str()), const_cast<char*>(body.c_str()), options))
+
+    std::vector<std::string> v;
+    for (int i = 0; i < r->result_length; i++)
+      v.push_back(r->result[i]);
```
Comment (Aschen) [inline src/document.cpp]: You can initialize a `std::vector<std::string>` directly from a `char**` if you know the length:

```
std::vector<std::string> v = std::vector<std::string>(r->result, r->result + r->result_length)
```

File: src/document.cpp [hunk 5]
```diff
@@ -17,221 +17,189 @@
...
+    KUZZLE_API(string_result, r, kuzzle_document_mcreate_or_replace(_document, const_cast<char*>(index.c_str()), const_cast<char*>(collection.c_str()), const_cast<char*>(body.c_str()), options))
+
+    std::string ret = r->result;
+    kuzzle_free_string_result(r);
+    return ret;
+  }
+
+  std::vector<std::string> Document::mDelete(const std::string& index, const std::string& collection, const std::vector<std::string>& ids, query_options *options) {
+    char **idsArray = new char *[ids.size()];
+    int i = 0;
```
Comment (Aschen) [inline src/document.cpp]: Useless variable creation in outer scope. A `size_t` could be used in the `for` for both array

File: src/document.cpp [hunk 6]
```diff
@@ -17,221 +17,189 @@
...
+
+    KUZZLE_API(
+      string_array_result,
+      r,
+      kuzzle_document_mdelete(_document, const_cast<char*>(index.c_str()), const_cast<char*>(collection.c_str()), idsArray, ids.size(), options),
+      delete[] idsArray)
+
+    std::vector<std::string> v;
+    for (int j = 0; j < r->result_length; j++)
+      v.push_back(r->result[j]);
```
Comment (Aschen) [inline src/document.cpp]: You can initialize a `std::vector<std::string>` directly from a `char**` if you know the length:

```
std::vector<std::string> v = std::vector<std::string>(r->result, r->result + r->result_length)
```

File: src/document.cpp [hunk 7]
```diff
@@ -17,221 +17,189 @@
...
+
+    kuzzle_free_string_array_result(r);
+    return v;
+  }
+
+  std::string Document::mGet(const std::string& index, const std::string& collection, const std::vector<std::string>& ids, query_options *options) {
+    char **idsArray = new char *[ids.size()];
+    int i = 0;
```
Comment (Aschen) [inline src/document.cpp]: Useless variable creation in outer scope. A `size_t` could be used in the `for` for both array

File: src/index.cpp [hunk 8]
```diff
@@ -16,106 +16,93 @@
...
+    KUZZLE_API(
+      string_array_result,
+      r,
+      kuzzle_index_mdelete(_index, indexesArray, indexes.size(), options),
+      delete[] indexesArray
+    )
+
+    std::vector<std::string> v;
+    for (int i = 0; i < r->result_length; i++)
+      v.push_back(r->result[i]);
```
Comment (Aschen) [inline src/index.cpp]: You can initialize a `std::vector<std::string>` directly from a `char**` if you know the length:

```
std::vector<std::string> v = std::vector<std::string>(r->result, r->result + r->result_length)
```

File: src/index.cpp [hunk 9]
```diff
@@ -16,106 +16,93 @@
...
+    kuzzle_free_bool_result(r);
+    return ret;
+  }
+
+  std::vector<std::string> Index::list(query_options *options) {
+    KUZZLE_API(string_array_result, r, kuzzle_index_list(_index, options))
+
+    std::vector<std::string> v;
+    for (int i = 0; i < r->result_length; i++)
+      v.push_back(r->result[i]);
```
Comment (Aschen) [inline src/index.cpp]: You can initialize a `std::vector<std::string>` directly from a `char**` if you know the length:

```
std::vector<std::string> v = std::vector<std::string>(r->result, r->result + r->result_length)
```

File: src/realtime.cpp [hunk 10]
```diff
@@ -55,16 +57,19 @@ namespace kuzzleio {
...
-    subscribe_result *r = kuzzle_realtime_subscribe(_realtime, const_cast<char*>(index.c_str()), const_cast<char*>(collection.c_str()),  const_cast<char*>(body.c_str()), call_subscribe_cb, this, options);
-    if (r->error != nullptr)
-        throwExceptionFromStatus(r);
+    KUZZLE_API(
+      subscribe_result,
+      r,
+      kuzzle_realtime_subscribe(_realtime, const_cast<char*>(index.c_str()), const_cast<char*>(collection.c_str()),  const_cast<char*>(body.c_str()), call_subscribe_cb, this, options))
 
     std::string roomId = r->room;
     std::string channel = r->channel;
```
Comment (Aschen) [inline src/realtime.cpp]: Unnecessary variable creation

File: src/websocket.cpp [hunk 11]
```diff
@@ -1,149 +1,148 @@
...
-      }
-    }
-
-    void trigger_websocket_notification_listener(notification_result* result, void* data) {
-      NotificationListener* listener = static_cast<WebSocket*>(data)->getNotificationListener(result->room_id);
-      if (listener) {
-        (*listener)(result);
+  void trigger_websocket_event_listener(int event, char* res, void* data) {
+    std::list<EventListener*> listeners = static_cast<WebSocket*>(data)->getListeners(event);
+    if (listeners.size()) {
```
Comment (Aschen) [inline src/websocket.cpp]: Why checking if the list is empty ? The for loop do the tricks if there is no element inside

File: src/websocket.cpp [hunk 12]
```diff
@@ -1,149 +1,148 @@
...
-      return res;
-    }
-
-    std::string WebSocket::close() {
-      const char* res = kuzzle_websocket_close(this->_web_socket);
-      if (res) {
-        return std::string(kuzzle_websocket_close(this->_web_socket));
+  void trigger_websocket_once(int event, char* res, void* data) {
+    std::list<EventListener*> listeners = static_cast<WebSocket*>(data)->getOnceListeners(event);
+    if (listeners.size()) {
```
Comment (Aschen) [inline src/websocket.cpp]: Unnecessary check

Comment (Aschen) [inline src/websocket.cpp]: Unnecessary check

File: src/index.cpp (hunk 2 above)
Comment (scottinet) [reply src/index.cpp]: I did not change that code, nor did I write it. I merely fixed the indentation: https://github.com/kuzzleio/sdk-cpp/blob/1-dev/src/index.cpp#L51

But yes, I'm fixing it while I'm at it.

File: src/index.cpp [hunk 13]
```diff
@@ -16,106 +16,93 @@
...
-            throwExceptionFromStatus(r);
-        bool ret = r->result;
-        kuzzle_free_bool_result(r);
-        return ret;
+  Index::Index(Kuzzle* kuzzle) {
+    _index = new kuzzle_index();
+    kuzzle_new_index(_index, kuzzle->_kuzzle);
+  }
+
+  Index::Index(Kuzzle* kuzzle, kuzzle_index *index) {
```
Comment (etrousset) [inline src/index.cpp]: If this is a copy constructor, why would we re init the struct with kuzzle_new_index() ?
And this brings me to what is tu purpose of this constructor?
If the user prepared the struct prior to calling the Index controller constructor the the configuration is lost, isn't it?

File: src/websocket.cpp (hunk 11 above)
Comment (scottinet) [reply src/websocket.cpp]: Dunno, not my code: https://github.com/kuzzleio/sdk-cpp/blob/1-dev/src/websocket.cpp#L13

Removing the useless check anyway.

File: src/index.cpp (hunk 13 above)
Comment (etrousset) [reply src/index.cpp]: Also, it looks like index content would be reallocated without out be freed

Comment (scottinet) [reply src/index.cpp]: Dunno, this is not my code: https://github.com/kuzzleio/sdk-cpp/blob/1-dev/src/index.cpp#L25

Comment (scottinet) [reply src/index.cpp]: I propose that we fix that issue in another PR. Can you write a new issue?

Comment (etrousset) [reply src/index.cpp]: :+1:

**Pull request merged**